src/astronomer/flask_appbuilder/security.py

```diff
@@ -310,15 +290,30 @@ def reload_jwt_signing_cert(self):
         """
         Reload (or load) the JWT signing cert from disk if the file has been modified.
         """
-        stat = os.stat(self.jwt_signing_cert_path)
-        if stat.st_mtime_ns > self.jwt_signing_cert_mtime:
-            log.info('Loading Astronomer JWT signing cert from %s', self.jwt_signing_cert_path)
-            with open(self.jwt_signing_cert_path, 'rb') as fh:
-                self.jwt_signing_cert = jwk.JWK.from_pem(fh.read())
-                # This does a second stat, but only when changed, and ensures
-                # that the time we record matches _exactly_ the time of the
-                # file we opened.
-                self.jwt_signing_cert_mtime = os.fstat(fh.fileno()).st_mtime_ns
+        try:
+            stat = os.stat(self.jwt_signing_cert_path)
+            if stat.st_mtime_ns > self.jwt_signing_cert_mtime:
+                log.info(
+                    "Loading Astronomer JWT signing cert from %s",
+                    self.jwt_signing_cert_path,
+                )
+                with open(self.jwt_signing_cert_path, "rb") as fh:
+                    self.jwt_signing_cert = jwk.JWK.from_pem(fh.read())
+                    # This does a second stat, but only when changed, and ensures
+                    # that the time we record matches _exactly_ the time of the
+                    # file we opened.
+                    self.jwt_signing_cert_mtime = os.fstat(fh.fileno()).st_mtime_ns
+        except FileNotFoundError:
+            from airflow.configuration import conf
+
+            # Example: http://houston-astronomer:8871/v1/.well-known/jwks.json
+            houston_url = conf.get("astronomer", "houston_jwk_url")
+            httprequest = Request(
+                houston_url, method="GET", headers={"Accept": "application/json"}
+            )
+            with urlopen(httprequest) as response:
+                key = response.read().decode()
+            self.jwt_signing_cert = jwk.JWK.from_json(key=key)
```

Comment:
This gets called on _every_ request that hits AIrflow (except to `/static/*`), so we'll want to do _some_ form of caching in process so that we don't need every single request to Airflow to hit Houston - that would slow down the webserver and put a lot of unnecessary load on Houston.

Reply:
@ashb what do you think about adding lru_cache with timeout?
```
def lru_cache_time(max_age, maxsize=128, typed=False):
    """
    Adds time aware caching to lru_cache
    """
    def wrapper(fn):
        @functools.lru_cache(maxsize=maxsize, typed=typed)
        def _new(*args, __time_salt, **kwargs):
            return fn(*args, **kwargs)
            
        @functools.wraps(fn)
        def _wrapped(*args, **kwargs):
            return _new(*args, **kwargs, __time_salt=int(time.time() / max_age))
            
        return _wrapped
    return wrapper


```

and replace code with:
```
        except FileNotFoundError:
            key = self._get_jwt_key_from_houston()
            self.jwt_signing_cert = jwk.JWK.from_json(key=key)


    @time_cache(10)
    def _get_jwt_key_from_houston(self):
        from airflow.configuration import conf

        # Example: http://houston-astronomer:8871/v1/.well-known/jwks.json
        houston_url = conf.get("astronomer", "houston_jwk_url")
        httprequest = Request(
            houston_url, method="GET", headers={"Accept": "application/json"}
        )
        with urlopen(httprequest) as response:
            key = response.read().decode()
        return key
```

Reply:
I wouldn't suggest using lru_cache for this, at least not like that as it would mean that we (unnecessarily) keep 128 old values.

Change `time.time()` to `time.monotonic()`, hard-code the maxsize to 1, and probably default timeout to 60 or 300 seconds, oh and make that timeout a setting from the Airflow config  (with a `fallback=60` say) then 👍🏻

Reply:
Oh, and we may as well cache the JWK object, not just the response.

Reply:
@ashb fixed.